lib/iterator.js

```diff
@@ -0,0 +1,317 @@
+'use strict';
+
+const toIterator = base => {
+  if (base[Symbol.iterator]) {
+    return base[Symbol.iterator]();
+  } else if (base.next) {
+    return base;
+  } else {
+    throw new TypeError('Base is neither Iterator nor Iterable');
+  }
+};
+
+class Iterator {
+  constructor(base) {
+    this.base = toIterator(base);
+  }
+
+  [Symbol.iterator]() {
+    return this;
+  }
+
+  next() {
+    return this.base.next();
+  }
+
+  each(fn, thisArg) {
+    this.forEach(fn, thisArg);
+  }
+
+  forEach(fn, thisArg) {
+    for (const value of this) {
+      fn.call(thisArg, value);
+    }
+  }
+
+  every(predicate, thisArg) {
+    for (const value of this) {
+      if (!predicate.call(thisArg, value)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  find(predicate, thisArg) {
+    for (const value of this) {
+      if (predicate.call(thisArg, value)) {
+        return value;
+      }
+    }
+  }
+
+  includes(element) {
+    for (const value of this) {
+      if (value === element || (isNaN(value) && isNaN(element))) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  reduce(reducer, initialValue) {
+    let result = initialValue;
+
+    if (result === undefined) {
+      const next = this.next();
+      if (next.done) {
+        throw new TypeError(
+          'Reduce of consumed iterator with no initial value'
+        );
+      }
+      result = next.value;
+    }
+
+    for (const value of this) {
+      result = reducer(result, value);
+    }
+    return result;
+  }
+
+  some(predicate, thisArg) {
+    for (const value of this) {
+      if (predicate.call(thisArg, value)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  collectTo(CollectionClass) {
+    return new CollectionClass(this);
+  }
+
+  collectWith(obj, collector) {
```

Comment:
@lundibundi you were the one who requested this particular implementation.

Reply:
Yeah, it's just `forEach` with a bit of state. I suggested it to be able to do
`iter(...).collectWith(new Struct(), (el, obj) => obj.add(el))`
Instead of having 
```
const obj = new Struct()
iter(...).forEach(el => obj.add(el))
```
Or
```
iter(...).reduce((obj, el) => (obj.add(el), obj), new Struct())
```

Reply:
@lundibundi in your particular example `Struct` constructor would probably accept an optional iterable to support `new Struct(elementsArray)`, so you'd be able to do `iter(elementsArray).collectTo(Struct)`, fwiw. That said, I understand that it's not always possible generally (e.g., you might need to collect data to a plain object). In this case, what exactly do find unergonomic in `reduce()`? Just the order of arguments or am I missing something?

Reply:
Yeah, basically order and the need to return the object every time even though it doesn't change. Also given the simplicity of implementation I suggested to just have it.
(Yeah, I assumed that it doesn't have such constructor/additional processing needed)

Reply:
@lundibundi yeah, I overlooked the fact that you don't need to return the same object you mutate imperatively anyway with `collectWith`, thanks. Should we maybe keep the reducer arguments order the same though?

Reply:
No strong opinion, though I assume they are this way in reduce due to the optionality of initialItem and here we have it as required. Also I kind of like when fuction comes last as it's clearer imo. Though if you or @nechaido want to keep same order as in reduce that's fine with me too.

Reply:
@lundibundi I agree with this. I was talking about the reducer signature, not the method one.

Reply:
Changed.